### ISSUE TITLE
Does snowman handle intrinsic functions?

### ISSUE THREAD

**Issue by hlide** [yegord/snowman#37]
Most compilers allow to insert special instructions through intrinsic functions. This is a way to avoid having .asm files and a good way to help the compiler to be aware of which registers are involved in a intrinsic function and to clobber the necessary ones.

It would be interesting to allow snowman to issue specific intrinsic function instead of inline assembly so it can link the terms used as the arguments with previous statements and the result of the function to future statements, what inline assembly statements are unable to do.

Of course those intrinsic functions are specific to an architecture and do not obey the same ABI rules as standard functions. They must be seen as *user* named N-ary operators.

**Comment by nihilus** [comment]
Agree, some architectures like the AVR8 have a special 4-bit swap of an 8-bit register. I guess this is a typical example where in intrisic would be nice as well as for some MIPS ASE instructions.

**Comment by yegord** [comment]
On Fri, Aug 07, 2015 at 08:48:31AM -0700, hlide wrote:
> It would be interesting to allow snowman to issue specific intrinsic
> function instead of inline assembly so it can link the terms used as
> the arguments with previous statements and the result of the function
> to future statements, what inline assembly statements are unable to
> do.

I suggest that you design the interface for intrinsics that you want to
have: what you are going to write in InstructionAnalyzer to use them,
how you want to tell the decompiler what is the semantics of the
intrinsics, what the user is going to see in the LikeC code. Code
samples would be great. And then we can discuss.

-- 
Yegor Derevenets

**Comment by hlide** [comment]
## `HADDPS xmm1, xmm2/m128`

### Intel C/C++ Compiler Intrinsic Equivalent

`__m128 _mm_hadd_ps (__m128 a, __m128 b);`

### snowman c++:
`xmm0_3 = _mm_hadd_ps(xmm2_1, xmm3_2);`

**Comment by hlide** [comment]
Intrinsic function looks like an ordinary function outside but it acts more like an N-ary operator inside:

It is black box and there is no algorithm behind.

*XxxInstructionAnalyzer*:
```
case I_xxx: {
    _(intrinsic_function(operand(0), "__builtin_xxx", operand(1), operand(2), operand(3)));
    break;
}  
```
asm:
```asm
    xxx %r0, %r1, %r2, $5
```

*snowman* c++:
```c++
c = __builtin_xxx(a, b, 5);
```
where *a*, *b* and *c* are terms;

*AllegrexInstructionAnalyzer*:
```
case I_ROTR: {
    _(intrinsic_function(gpr(0), "__allegrex_rotr", gpr(1), imm(2)));
    break;
}  
```
asm:
```asm
    rotr %v0, %a0, $4
```

*snowman* c++:
```c++
v0_2 = __allegrex_rotr(a0_1, 4);
```

**Comment by yegord** [comment]
On Fri, Aug 07, 2015 at 02:41:14PM -0700, hlide wrote:
> It is black box and there is no algorithm behind.

How should DataflowAnalyzer do constant folding for it? Or it should
not? How should TypeAnalyzer infer types of the arguments and of the
result? Or it should not?

Also, in LikeC representation every function called must have a
declaration, with the types of the arguments and the result. Do you plan
to hand-code the declaration for every built-in?

-- 
Yegor Derevenets

**Comment by hlide** [comment]
1) DataflowAnalyzer: if it is a black box, constant folding is not possible. We could optionally add semantics to an *intrisic function* to allow constant folding but I do not see the point to do so because it would mean all the input terms are constants and so why the compiler did not optimize them out?

2) TypeAnalyzer: good point. But I think you already have that piece of information through input and output terms of the *intrinsic function*. You do not really more.

3) *intrinsic function* is not a function. It does not obey ABI rules. for instance in MIPS, function argument are expected to be in registers *%a0*, *%a1*, and so on, while *intrinsic function* acts more like a N-ary operation regarding with arguments and result (it can accept anything)

**Comment by yegord** [comment]
On Sat, Aug 08, 2015 at 02:48:41AM -0700, hlide wrote:
> We could optionally add semantics to an *intrisic function*
How? Give the actual expression to the constructor of the intrinsic
function term?

> it would mean all the input terms are constants and so why the
> compiler did not optimize them out?
I do not understand what and how it should optimize out.

For example, the decompiler can currently infer from `eax = eax |
0xffffffff` that eax will become 0xffffffff, although eax is not a
constant. Do you want similar things to work for intrinsic functions?
Or you are fine with the decompiler assuming that they return anything?

> 2) TypeAnalyzer: good point. But I think you already have that piece
> of information through input and output terms of the *intrinsic
> function*.
Okay.

> 3) *intrinsic function* is not a function. It does not obey ABI rules.
> for instance in MIPS, function argument are expected to be in
> registers *%a0*, *%a1*, and so on, while *intrinsic function* acts
> more like a N-ary operation regarding with arguments and result (it
> can accept anything)
So, an intrinsic function should be a special LikeC expression. Okay.

What is the overall goal? Just to give the user something like

    r1 = swap_halfs(r2)

instead of

    r1 = (r2 << 4) | (r2 >> 4)

?

-- 
Yegor Derevenets

**Comment by hlide** [comment]
1) supposedly that we add semantics, it could be something like that:
```c++
case I_ROTR: {
    AllegrexExpressionFactoryCallback body(factory, program->createBasicBlock(), instruction);
    auto gpr0 = gpr(0);
    auto gpr1 = gpr(1);
    auto imm2 = imm(2);
    _(intrinsic_function_callback(body, gpr0, "__allegrex_rotr", gpr1, imm2) [
        _[gpr0 ^= ((unsigned_(gpr1) >> imm2) | (gpr1 << (constant(32) - imm2))];
    ]);
    break;
} 
```
unless the result becomes a constant, we keep as c++ output:
```c++
a = __allegrex_rotr(b, c);
```

But again the C or C++ compiler which generated a *ROTR* would have already done the constant folding and this instruction would have been replaced with a constant, so the more I think, the less having semantics added for an *instrinsic function* sounds useful. And your example with `eax = eax | 0xffffffff` should not have occurred as the C/C++ compiler should have it optimized out with issuing at most `eax = 0xffffffff`instead.

Instead of transforming them into an obfuscating expression (think about *LWL* or *LWR* examples), it is preferable to keep the original intent through a named "lambda" function.

2) On the second hand, it may be interesting to have some possible transformations on *intrisic functions*. For MIPS, the pseudo instruction *ULW* is in fact two instructions *LWL* and *LWR* with specific offsets 3 and 0 (which case is almost 99%). Both of those instructions may turn into a single *intrinsic function* like `rd = __unaligned_load(rb)` which is certainly more readable that the two very obfuscating expressions for each *LWL* and *LWR*. But that is another story and anyway something specific to an architecture.
 
3) what is the goal? retain the original specific instruction especially when it may be turned into an obfuscating expression or you want it to be recompilable through an intrinsic function or even a macro with a C/C++ compiler. Or simply to make it more readable, still close to the original one because the original source also uses intrinsic functions (__allegrex_xxx, __builtin_xx, and so on in gcc).

**Comment by yegord** [comment]
On Sat, Aug 08, 2015 at 06:41:34AM -0700, hlide wrote:
> 1) supposedly that we add semantics, it could be something like that:
> ```c++
> case I_ROTR: {
>     AllegrexExpressionFactoryCallback body(factory, program->createBasicBlock(), instruction);
>     auto gpr0 = gpr(0);
>     auto gpr1 = gpr(1);
>     auto imm2 = imm(2);
>     _(intrinsic_function_callback(body, gpr0, "__allegrex_rotr", gpr1, imm2) [
>         _[gpr0 ^= ((unsigned_(gpr1) >> imm2) | (gpr1 << (constant(32) - imm2))];
>     ]);
>     break;
> } 

So, you propose to define the semantics of an intrinsic as a control
flow graph, thus making the control flow graph a multilevel graph. To me
this look like a heavy complication for a not too big profit.

> And your example with `eax = eax | 0xffffffff` should not have
> occurred as the C/C++ compiler should have it optimized out with
> issuing at most `eax = 0xffffffff`instead.

It was actually you who posted this example:
https://github.com/yegord/snowman/issues/27#issuecomment-123836626

> 3) what is the goal? retain the original specific instruction
> especially when it may be turned into an obfuscating expression or you
> want it to be recompilable through an intrinsic function or even a
> macro with a C/C++ compiler. Or simply to make it more readable, still
> close to the original one because the original source also uses
> intrinsic functions (__allegrex_xxx, __builtin_xx, and so on in gcc).

Okay. I can buy the readability argument. However, it is currently
possible to see where a certain complicated expression comes from by a
single click on it. I am not convinced that duplicating the instruction
(as an intrinsic function) in the LikeC code gives substantial benefits
here.

Summing up, I do not think that this feature is very necessary, however,
I will not stop anybody from implementing it (one needs a new kind of
Term, support for it in all the analyses, a new kind of LikeC
expression) and sending me a patch with an example where it makes life
better.

-- 
Yegor Derevenets

**Comment by hlide** [comment]
Well, the argument about `eax = eax | 0xffffffff` (`xor %eax,-1`) is still not viable because it is not meant to be turned into a *intrinsic function* anyway. You're looking for false assumption and this *instrinsic function* is clearly a lack for some goals probably different than yours. I want to focus my efforts on Allegrex architecture implementation, not on common or core implementation I am not fluent with.

I don't want  a complicated way to create an *intrinsic function*:

1) black box, no semantics, no constant folding.
2) types are extracted from output and input terms, as such.
3) implemented intrinsic functions with name, output and inputs terms are done in ArchitectureAnalyzer.
4) using them instead of inline assembly expression may help to debug the c++ output of an instruction by retaining links between referenced registers or memory slots in the chains of expressions instead of breaking them.
5) simple c++ output: *[*output =*]* name(*[*input1, ..., inputN*]*); 
6) quick implementations for SSE and AVX instructions (seem missing in the actual implementation)

Obviously they can be composed:
```c++
xmm0_2 = _mm_add_ps(xmm1_1, _mm_mul_ps(xmm2_1));
```

**Comment by hlide** [comment]
By the way, Allegrex has instructions for complex maths like *sin*, *cos*, *exp*, *log*, *sqrt*, *rsqrt* and so on. How I'm suppose to translate them?

**Comment by yegord** [comment]
On Sat, Aug 08, 2015 at 12:10:00PM -0700, hlide wrote:
> By the way, Allegrex has instructions for complex maths like *sin*,
> *cos*, *exp*, *log*, *sqrt*, *rsqrt* and so on. How I'm suppose to
> translate them?

It is a good point. My first choice would be: add new kinds of unary and
binary operators on IR level (you anyway will need things like
floating-point addition, multiplication, negation), hardcode/create on
demand function declarations on LikeC level and generate calls to them
for these IR operators.

The second possible solution would be intrinsics as you describe them.

Actually, the difference between the approaches is mainly in the naming.
And in the approach with intrinsics sin/cos become in LikeC not
functions, but some special operators, which I find a bit weird.

Concerning _mm_mul_ps, __allegrex_rotr, and similar, my plan would be,
to describe their semantics in terms of basic operations like
floating-point multiplications, or bit shifts. The advantage of this
approach, is that existing analyses just continue working (assuming, you
have implemented basic floating-point operations). The disadvantage, of
course, is, that you have to define the semantics. :)

I am just saying, what I would do, but I am a human, I can be wrong.

-- 
Yegor Derevenets

**Comment by hlide** [comment]
well that's exactly what intrinsic "functions" are for C/C++ compilers like gcc or visual studio: they look like functions but they are not functions in the usual way.

**Comment by hlide** [comment]
Just give me good indications how to add this *generic intrinsic function* in the IR. So far, I may need up to 4 input terms

**Comment by hlide** [comment]
closed to bad clic

**Comment by yegord** [comment]
On Sat, Aug 08, 2015 at 01:48:47PM -0700, hlide wrote:
> Just give me good indications how to add this *generic intrinsic
> function* in the IR. So far, I may need up to 4 input terms 

Make a new subclass of Term, storing, say, an std::vector of argument
terms. (One could possibly use std::array instead, to save on
allocations.) Generate an instance of it for some instruction, run
decompilation, and fix what breaks: I have plenty of assertions. :)

-- 
Yegor Derevenets

**Comment by yegord** [comment]
https://github.com/yegord/snowman/commit/f96ec64ff5d5358e006a39937c0e706264f2dc90 — this is how you remove a Term subclass. Adding is just the opposite.

**Comment by yegord** [comment]
I have added `likec::UndeclaredIdentifier` which makes generating intrinsic calls in LikeC fairly easy: https://github.com/yegord/snowman/blob/master/src/nc/core/ir/cgen/DefinitionGenerator.cpp#L1019.
I do not plan to do anything else on the topic.
Which new terms to add and what semantics to give to them — it is up to you.

**Comment by hlide** [comment]
Can you give an simple example how to use it and the resulting C++ output?

**Comment by nihilus** [comment]
Agree with @hlide, an example would be good I dont plan to implement unaliged double word load / store for mips64

**Comment by yegord** [comment]
An example: https://github.com/yegord/snowman/blob/master/src/nc/core/ir/cgen/DefinitionGenerator.cpp#L1001

It generates, for example the `__zero_stack_offset()` and `__return_address()` subexpressions in here (tests/hand-made/001_hello_world example):
```
    rsp1 = (void*)((int64_t)__zero_stack_offset() + 8);
    fun_4003f0(main, __return_address(), rsp1, __libc_csu_init, __libc_csu_fini, rdx2, ((uint64_t)rsp1 & 0xfffffffffffffff0) - 8 - 8, rax3);
```
To generate intrinsic calls with arguments,
1) when creating `likec::FunctionPointerType`, add types of arguments to it, see `addArgumentType` method,
2) when creating `likec::CallOperator`, add the arguments to it, see `addArgument` method.

**Comment by hlide** [comment]
I'm a little confused by the term of *UndeclaredIdentifier*. I would have expect for it to be *UndefinedIdentifier*, that is, a declared identifier (name is given) but with no definition (no implementation). Unless you mean an identifier with no type (no signature)?

**Comment by yegord** [comment]
It is not declared, in the sense that it does not have a declaration, unlike, say, VariableIdentifier or FunctionIdentifier. However, it has a type. The idea is, if you want to call a function which you do not want to declare, create an UndeclaredIdentifier, give it a pointer to a function type, and use it in a call expression.

**Comment by hlide** [comment]
Well the issue I have is I am working on IR level in *AllegrexInstructionAnalyzer* so I do not exactly how I can proceed since there is no stuff I can see to declare it with a name at this level.

**Comment by yegord** [comment]
If you want to generate some intrinsics in LikeC, you need some IR to generate them from.
So, you need to introduce a new intrinsic kind (see `ir::Intrinsic` class), unary operator kind (see `ir::UnaryOperator` class), binary operator kind (see `ir::BinaryOperator` class), or just a new `ir::Term` subclass, if you want something really special.
In `InstructionAnalyzer` you start creating terms of this new special kind.
Then, you extend `DefinitionGenerator` to generate LikeC intrinsics (or whatever else you may want) from them.

**Comment by hlide** [comment]
Since those intrinsic functions are architecture-dependent, I guess I cannot introduce any new kind whatever class is chosen. I am aware of those *xxx::USER* but there is no relative stuff to register new user kind in those base classes. So it sounds like I might need to derive one of those classes to make them aware of a specific architecture - as we need to put outside the place where the originals are.

The more I think, the more we need is:

1) when output in a C++ page, it must be like a named function with a result or not, with arguments or not.
2) we can add full or partial implementation to state some relations between result and arguments (like setting the conditional flags to help the conditional jumps) but we do not want to see them in the C++ output.

**Comment by yegord** [comment]
`xxx::USER` is a leftover from five years ago, and actually should be removed.

You can start with something like
```c++
class IntrinsicCall: public ir::Term {
    QLatin1String name_;
    std::vector<std::unique_ptr<ir::Term>> operands_;

public:
    IntrinsicCall(QLatin1String name, std::vector<std::unique_ptr<ir::Term>> operands): Term(INTRINSIC_CALL), name_(std::move(name)), operands_(std::move(operands)) {}

    const QLatin1String &name() const { return name_; }
    const std::vector<std::unique_ptr<ir::Term>> &operands() const { return operands_; }
};
```
Dataflow-wise, it can be equivalent to Intrinsic::UNDEFINED's behavior.
In DefinitionGenerator this translates to LikeC as a call to UndeclaredIdentifier with the name being the intrinsic's name with the type being a function pointer with the result type that has been computed by the decompiler, with the argument types that have been computed by the decompiler.

**Comment by nihilus** [comment]
This is a typical example where I'd wanna use the intrinsic:
```
    int32_t v22;

    __asm__("sdl $a1, 0x68($sp)");
    __asm__("sdr $a1, 0x6f($sp)");
    v2 = gp;
```
I'll try the suggestion of @yegord when I've implemented more 64-bit instructions.

**Comment by yegord** [comment]
And what do you want to get in the end?

**Comment by nihilus** [comment]
@yegord I would want something like an assignment  of  0x68($sp) and 0x6f($sp) to register '$a1'... I am sure that @hlide can give more complex and better examples than this one.

**Comment by hlide** [comment]
well in this case we could have a pattern to help to merge those two instructions into one intrinsic function (unaligned store 64-bit word operation).

Two possibilities:

1) disassembler merges those two instructions into `usd $a1, 0x68($sp)` then we turn it into `int64_t a1 = __builtin_mips_usd(stack_pointer() + 0x68);`. This function allows unaligned memory access whereas `int64_t a1 = stack_pointer()[0x68];` assumes an aligned memory address (at least for MIPS).

2) those instructions are transformed into intrinsic functions and a pattern analyzer will merge them into one intrinsic function:
```c++
...
int64_t a1_2 = __builtin_mips_sdl(a1_1, stack_pointer() + 0x68);
int64_t a1_3 = __builtin_mips_sdr(a1_2, stack_pointer() + 0x6F);
...
```
which gives us:
```c++
...
int64_t a1_3 = __builtin_mips_sdr(__builtin_mips_sdl(a1_1, stack_pointer() + 0x68), stack_pointer() + 0x6F);
...
```

since the bases are the same and xoring offsets 0x68 and 0x6F gives 0x07, we can merge those two intrinsic functions into:

```c++
...
int64_t a1_3 = __builtin_mips_usd(a1_1, stack_pointer() + 0x68);
...
```
The final offset is the lesser in both Little/Big-endian cases.

For 32-bit word version, xoring the offsets should give 0x03.

**Comment by nihilus** [comment]
@hlide That'd be neat.
